### DISCUSSION TITLE
Up-to-date testing guidelines/helpers?

### DISCUSSION THREAD

**Discussion by KholdStare** [wevm/wagmi#1762] [Q&A]
WAGMI evolves quite quickly, and I've read numerous discussions/issues etc regarding mocking and testing.

One thread mentioned this example repo: https://github.com/tmm/testing-wagmi . It is 5 months old now, and may be out of date. It also has a _lot_ of boilerplate to set everything up.

We need a way to thoroughly unittest our components for interactions between all the different states, and there are a lot of them - There's preparedContractWrite's `isLoading`, `isFetching`, `isIdle`, etc, then the actual contractWrite has its own things like `writeAsync` may or may not be present. We want to unit test these scenarios and make sure our React components don't end up glitching on some of these intermediate states.

I don't think running this alongside `anvil` or some external testing blockchain makes sense for unit tests, because it is not possible to consistently and deterministically step through all the state transitions that can happen. There needs to be a way of mocking out the wagmi hooks.

I've also considered mocking out https://github.com/TanStack/query `useQuery` to somehow drive the hooks along. Are there any plans to support this, or ideas on how to implement something like this? You have all this rich type information from the ABI, it would be great to have testing/mocking facilities that are also type-safe.

**Comment by lefuncq** [comment]
Do you have any follow-up on that request?
I still haven't figured out how to properly test components that use Wagmi without having a local chain running on the side.
If, at least, we could mock wagmi's named exports within our tests, this would allow us to deterministically test different flows of our apps.

**Comment by t0rbik** [comment]
Tbh i would be curious to look at any OS repo with testing of component with wagmi hook via local chain, so if there are any out there share please

**Comment by llllvvuu** [comment]
I run a local blockchain. But TBH, another way is to use the [separation of responsibilities pattern](https://sairys.medium.com/react-separating-responsibilities-using-hooks-b9c90dbb3ab9). Then you can test business logic re: state without having the state, and your integration tests don't need to cover as much ground as per ["Boundaries"](https://www.destroyallsoftware.com/talks/boundaries).
